pyomo/core/expr/compare.py

```diff
@@ -185,3 +189,79 @@ def compare_expressions(expr1, expr2, include_named_exprs=True):
     except PyomoException:
         res = False
     return res
+
+
+def assertExpressionsEqual(test, a, b, include_named_exprs=True):
+    """unittest-based assertion for comparing expressions
+
+    This converts the expressions `a` and `b` into prefix notation and
+    then compares the resulting lists.
+
+    Parameters
+    ----------
+    test: unittest.TestCase
+        The unittest `TestCase` class that is performing the test.
+
+    a: ExpressionBase or native type
+
+    b: ExpressionBase or native type
+
+    include_named_exprs: bool
+       If True (the default), the comparison expands all named
+       expressions when generating the prefix notation
+    """
+    prefix_a = convert_expression_to_prefix_notation(a, include_named_exprs)
+    prefix_b = convert_expression_to_prefix_notation(b, include_named_exprs)
+    try:
+        test.assertEqual(len(prefix_a), len(prefix_b))
+        for _a, _b in zip(prefix_a, prefix_b):
+            test.assertIs(_a.__class__, _b.__class__)
+            test.assertEqual(_a, _b)
+    except (PyomoException, AssertionError):
+        test.fail(f"Expressions not equal:\n\t"
+                  f"{tostr(prefix_a)}\n\t!=\n\t{tostr(prefix_b)}")
+
+
+def assertExpressionsStructurallyEqual(test, a, b, include_named_exprs=True):
+    """unittest-based assertion for comparing expressions
+
+    This converts the expressions `a` and `b` into prefix notation and
+    then compares the resulting lists.  Operators and (non-native type)
+    leaf nodes in the prefix epresentation are converted to strings
+    before comparing (so that things like variables can be compared
+    across clones or pickles)
+
+    Parameters
+    ----------
+    test: unittest.TestCase
+        The unittest `TestCase` class that is performing the test.
+
+    a: ExpressionBase or native type
+
+    b: ExpressionBase or native type
+
+    include_named_exprs: bool
+       If True (the default), the comparison expands all named
+       expressions when generating the prefix notation
+
+    """
+    prefix_a = convert_expression_to_prefix_notation(a, include_named_exprs)
+    prefix_b = convert_expression_to_prefix_notation(b, include_named_exprs)
+    # Convert leaf nodes and operators to their string equivalents
+    for prefix in (prefix_a, prefix_b):
+        for i, v in enumerate(prefix):
+            if type(v) in native_types:
+                continue
+            if type(v) is tuple:
+                if len(v) == 3:
+                    prefix[i] = v[:2] + (str(v[2]),)
+                continue
+            prefix[i] = str(v)
+    try:
+        test.assertEqual(len(prefix_a), len(prefix_b))
+        for _a, _b in zip(prefix_a, prefix_b):
+            test.assertIs(_a.__class__, _b.__class__)
```

Comment:
Do you still need to check the class--aren't they all strings? Oh, wait, except for native types, right?

Reply:
Yes.  This is comparing elements of prefix, which may be native types, tuples, strings, etc.